# 06 — Consensus Backtest: Does Mutual-Fund Consensus Predict Returns?

**Question:** do stocks held by *more* mutual-fund schemes (`holders_total` in `consensus_panel`)
earn higher forward returns than stocks held by fewer schemes?

**Method (SQL-only against live Postgres — all business logic lives in the SQL statements below,
cells only orchestrate, count, and plot):**

1. For every `(isin, qtr)` row in `consensus_panel`, bucket the stock into a **decile of
   `holders_total` within its quarter** (`ntile(10) OVER (PARTITION BY qtr ORDER BY holders_total)`).
2. Compute the **t → t+4-quarter forward return per ISIN** from `security_prices`
   (entry = last close on/before quarter start, look-back tolerance 45 days;
   exit = close nearest to quarter start + 12 months, tolerance ±45 days).
3. Subtract the **NIFTY SMALLCAP 250** forward return over the *identical* window
   (same entry/exit date rules on `index_prices`) → **forward excess return**.
4. Report mean/median forward excess return by decile + per-quarter diagnostics + plot.

**Honesty guards (all counts printed below):**
- *Survivorship:* the panel is built from historical `portfolio_snapshots`, so it **includes exited
  holdings** (stocks later dropped by schemes still appear in their historical quarters) — this
  mitigates, but does not eliminate, survivorship bias (delisted symbols still have price gaps
  post-Jul-2024 per the ingestion caveats).
- *Missing prices:* panel rows whose ISIN has **no entry price** (mostly debt / repo / G-Sec / junk
  ISINs, which have no equity price history) are excluded and counted.
- *Short histories:* rows with an entry price but **no observable exit ≥ ~4 forward quarters**
  (no close within ±45 days of quarter start + 12 months) are excluded and counted.
- *Overlapping quarters:* panel coverage is deep only for recent quarters (B2 backfill got
  static-HTML AMCs), so usable (t, t+4q) windows are limited — every quarter's row count is printed.

**Compliance note:** this is a descriptive data-quality/sanity analysis of public disclosure data —
not investment advice.

In [1]:
# [E3] Entry-mode configuration - fixes structural look-ahead bias.
#
# SEBI requires monthly portfolio disclosure within 10 calendar days of
# month end; most AMCs publish by the 8th-10th. A quarter's panel is only
# fully knowable after its LAST month is published:
#   earliest_safe_entry = qtr_end + 10 days (+1 day safety)
#                       = qtr_start + ~3 months + 11 days.
#
#   quarter_start       - OLD behaviour: enter at the quarter's first day;
#                         knowingly trades on information published later
#                         (look-ahead bias). Kept so both variants stay
#                         reproducible and the bias delta is measurable.
#   disclosure_aligned  - enter at qtr_end + 11 days; exit = entry + 12m.
ENTRY_MODE = "disclosure_aligned"   # or "quarter_start"
assert ENTRY_MODE in ("quarter_start", "disclosure_aligned")

ENTRY_SHIFT_SQL = (
    "INTERVAL '0 days'"
    if ENTRY_MODE == "quarter_start"
    else "INTERVAL '3 months' + INTERVAL '11 days'"  # qtr -> qtr_end + 11d
)
print(f"[E3] ENTRY_MODE={ENTRY_MODE} | entry anchor shift={ENTRY_SHIFT_SQL}")


[E3] ENTRY_MODE=disclosure_aligned | entry anchor shift=INTERVAL '3 months' + INTERVAL '11 days'


In [2]:
import logging
import sys
from pathlib import Path

import numpy as np
import pandas as pd
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
from sqlalchemy import create_engine, text

logging.basicConfig(level=logging.INFO, format="%(asctime)s | %(levelname)s | %(message)s",
                    stream=sys.stdout)
log = logging.getLogger("06_consensus_backtest")

REPO = Path.cwd()
while not (REPO / "AGENTS.md").exists() and REPO != REPO.parent:
    REPO = REPO.parent
REPORT_DIR = REPO / "data" / "reports" / "mutual_funds"
REPORT_DIR.mkdir(parents=True, exist_ok=True)

DB_URL = "postgresql+psycopg2://vlmrouter:vlmrouter@localhost:5432/mutual_funds"
eng = create_engine(DB_URL, pool_pre_ping=True)

def q(sql: str) -> pd.DataFrame:
    # run one SQL statement against live Postgres and return a DataFrame
    with eng.connect() as conn:
        return pd.read_sql(text(sql), conn)

BENCHMARK = "NIFTY SMALLCAP 250"
DEMO_ISIN, DEMO_NAME = "INE953O01021", "Sansera Engineering Limited"  # real stock held by small-cap schemes

with eng.connect() as conn:
    db_ok = conn.execute(text("SELECT version()")).scalar()
log.info("[STAGE 0] connected to Postgres: %s", db_ok.split(",")[0])
log.info("[STAGE 0] report dir: %s", REPORT_DIR)

2026-08-22 17:54:15,229 | INFO | [STAGE 0] connected to Postgres: PostgreSQL 18.3 (Debian 18.3-1.pgdg12+1) on aarch64-unknown-linux-gnu


2026-08-22 17:54:15,229 | INFO | [STAGE 0] report dir: /Users/vedaangchopra/all_data/complete_technical_work/all_projects_implemented/Financial Analytics Work/data/reports/mutual_funds


In [3]:
log.info("[STAGE 1] consensus_panel overview (SQL view only)")

panel_by_qtr = q("""
    SELECT qtr,
           COUNT(*)                        AS n_rows,
           COUNT(DISTINCT isin)            AS n_isins,
           MIN(holders_total)              AS min_holders,
           MAX(holders_total)              AS max_holders,
           SUM((holders_total >= 2)::int)  AS rows_multi_holder
    FROM consensus_panel
    GROUP BY qtr
    ORDER BY qtr
""")
panel_tot = q("SELECT COUNT(*) AS n_rows, COUNT(DISTINCT isin) AS n_isins FROM consensus_panel")

print(f"consensus_panel total rows : {panel_tot.loc[0, 'n_rows']:,}")
print(f"consensus_panel unique ISINs: {panel_tot.loc[0, 'n_isins']:,}")
print(f"quarters with panel data   : {panel_by_qtr.shape[0]}")
print("\nRows per quarter:")
print(panel_by_qtr.to_string(index=False))

2026-08-22 17:54:15,233 | INFO | [STAGE 1] consensus_panel overview (SQL view only)


consensus_panel total rows : 55,602
consensus_panel unique ISINs: 18,662
quarters with panel data   : 53

Rows per quarter:
       qtr  n_rows  n_isins  min_holders  max_holders  rows_multi_holder
2013-04-01      18       18            1            1                  0
2013-07-01      23       23            1            1                  0
2013-10-01      24       24            1            1                  0
2014-01-01      25       25            1            1                  0
2014-04-01      28       28            1            1                  0
2014-07-01      29       29            1            1                  0
2014-10-01      35       35            1            1                  0
2015-01-01      29       29            1            1                  0
2015-04-01      31       31            1            1                  0
2015-07-01      32       32            1            1                  0
2015-10-01      30       30            1            1                  0


In [4]:
log.info("[STAGE 2] per-ISIN t -> t+4q forward returns from security_prices (SQL lateral joins)")

STOCK_FWD_SQL = f"""
WITH
    -- Equity universe only: drop debt/G-Sec/SDL/repo/cash/CD/CP rows.
    -- These carry no equity price history and previously polluted the
    -- missing-price diagnostics (~5.1K of 18.2K rows in the first run).
    panel AS (
        SELECT *
        FROM consensus_panel
        WHERE isin LIKE 'INE%'
          AND NOT lower(coalesce(instrument_name, '')) ~
              '(government|g-?sec|treasury|t-bill|tbill|state development|sdl|reverse repo|treps|tri-party|cblo|mibor|bill discounting|certificate of deposit|commercial paper|^cash| cash$|margin)'
    ),
    px AS (
    SELECT
        cp.isin,
        cp.qtr,
        cp.holders_total,
        e.close      AS entry_close,
        e.trade_date AS entry_date,
        x.close      AS exit_close,
        x.trade_date AS exit_date
    FROM panel cp
    -- entry: last close on/before quarter start (tolerance: 45 days look-back)
    LEFT JOIN LATERAL (
        SELECT sp.close, sp.trade_date
        FROM security_prices sp
        WHERE sp.isin = cp.isin
          AND sp.trade_date <= cp.qtr + {ENTRY_SHIFT_SQL}
          AND sp.trade_date >= cp.qtr + {ENTRY_SHIFT_SQL} - INTERVAL '45 days'
        ORDER BY sp.trade_date DESC
        LIMIT 1
    ) e ON TRUE
    -- exit: close nearest to quarter start + 12 months (tolerance: +/- 45 days)
    LEFT JOIN LATERAL (
        SELECT sp.close, sp.trade_date
        FROM security_prices sp
        WHERE sp.isin = cp.isin
          AND sp.trade_date >= cp.qtr + {ENTRY_SHIFT_SQL} + INTERVAL '12 months' - INTERVAL '45 days'
          AND sp.trade_date <= cp.qtr + {ENTRY_SHIFT_SQL} + INTERVAL '12 months' + INTERVAL '45 days'
        ORDER BY ABS(sp.trade_date - (cp.qtr + {ENTRY_SHIFT_SQL} + INTERVAL '12 months')::date)
        LIMIT 1
    ) x ON TRUE
)
SELECT isin,
       qtr::date            AS qtr,
       holders_total,
       entry_date,
       entry_close::float8  AS entry_close,
       exit_date,
       exit_close::float8   AS exit_close
FROM px
"""

px = q(STOCK_FWD_SQL)
log.info("[STAGE 2] panel rows fetched: %d", len(px))

n_total = len(px)
no_entry = px["entry_close"].isna()
no_exit = px["entry_close"].notna() & px["exit_close"].isna()
ok = px["entry_close"].notna() & px["exit_close"].notna() & (px["entry_close"] > 0)

# after the equity-only filter every remaining ISIN is equity-looking
no_entry_eq = (no_entry & px["isin"].str.startswith("INE")).sum()
no_entry_other = (no_entry & ~px["isin"].str.startswith("INE")).sum()

print(f"[STAGE 2] panel rows (t observations)                     : {n_total:,}")
print(f"[STAGE 2] excluded - no entry price (missing price)       : {int(no_entry.sum()):,}"
      f"  (equity-looking INE ISINs: {int(no_entry_eq):,}; debt/repo/gsec/junk ISINs: {int(no_entry_other):,})")
print(f"[STAGE 2] excluded - entry ok but no ~4q forward price    : {int(no_exit.sum()):,}"
      f"  (short forward history / delisted gap)")
print(f"[STAGE 2] included - full entry + ~4q forward window      : {int(ok.sum()):,}")
assert int(no_entry.sum() + no_exit.sum() + ok.sum()) == n_total

stock_fwd = px.loc[ok, ["isin", "qtr", "holders_total", "entry_date", "entry_close",
                        "exit_date", "exit_close"]].copy()
stock_fwd["fwd_ret"] = stock_fwd["exit_close"] / stock_fwd["entry_close"] - 1.0
print("\n[STAGE 2] stock forward-return observations by quarter:")
print(stock_fwd.groupby("qtr").agg(n_obs=("isin", "size"), n_isins=("isin", "nunique")).to_string())

2026-08-22 17:54:15,334 | INFO | [STAGE 2] per-ISIN t -> t+4q forward returns from security_prices (SQL lateral joins)


2026-08-22 17:54:25,416 | INFO | [STAGE 2] panel rows fetched: 41477


[STAGE 2] panel rows (t observations)                     : 41,477
[STAGE 2] excluded - no entry price (missing price)       : 32,081  (equity-looking INE ISINs: 32,081; debt/repo/gsec/junk ISINs: 0)
[STAGE 2] excluded - entry ok but no ~4q forward price    : 3,653  (short forward history / delisted gap)
[STAGE 2] included - full entry + ~4q forward window      : 5,743

[STAGE 2] stock forward-return observations by quarter:
            n_obs  n_isins
qtr                       
2015-10-01     21       21
2016-01-01     25       25
2016-04-01     24       24
2016-07-01     20       20
2016-10-01     22       22
2017-01-01     19       19
2017-04-01     19       19
2017-07-01     19       19
2017-10-01     24       24
2018-01-01     25       25
2018-04-01     26       26
2018-07-01     26       26
2018-10-01     26       26
2019-01-01     24       24
2019-04-01     22       22
2020-04-01     23       23
2020-07-01     25       25
2020-10-01     55       55
2021-01-01     18       18
2021

In [5]:
log.info("[STAGE 3] NIFTY SMALLCAP 250 forward return over identical windows (SQL)")

IDX_FWD_SQL = f"""
WITH qtrs AS (
    SELECT DISTINCT qtr FROM consensus_panel
)
SELECT q.qtr::date AS qtr,
       e.close::float8      AS entry_close,
       e.trade_date         AS entry_date,
       x.close::float8      AS exit_close,
       x.trade_date         AS exit_date,
       (x.close / e.close - 1)::float8 AS fwd_ret
FROM qtrs q
JOIN LATERAL (
    SELECT ip.close, ip.trade_date
    FROM index_prices ip
    WHERE ip.index_symbol = '{BENCHMARK}'
      AND ip.trade_date <= q.qtr
    ORDER BY ip.trade_date DESC
    LIMIT 1
) e ON TRUE
JOIN LATERAL (
    SELECT ip.close, ip.trade_date
    FROM index_prices ip
    WHERE ip.index_symbol = '{BENCHMARK}'
      AND ip.trade_date >= q.qtr + INTERVAL '12 months' - INTERVAL '45 days'
      AND ip.trade_date <= q.qtr + INTERVAL '12 months' + INTERVAL '45 days'
    ORDER BY ABS(ip.trade_date - (q.qtr + INTERVAL '12 months')::date)
    LIMIT 1
) x ON TRUE
"""

idx_fwd = q(IDX_FWD_SQL)
log.info("[STAGE 3] %s windows computed for %d quarters", BENCHMARK, len(idx_fwd))

n_qtrs_panel = panel_by_qtr.shape[0]
n_qtrs_idx = idx_fwd["qtr"].nunique()
print(f"[STAGE 3] panel quarters                                  : {n_qtrs_panel}")
print(f"[STAGE 3] quarters with a full benchmark 4q window        : {n_qtrs_idx}")
print(f"[STAGE 3] quarters dropped (benchmark window incomplete)  : {n_qtrs_panel - n_qtrs_idx}")
print(f"\n[STAGE 3] {BENCHMARK} forward 12m return by quarter:")
print(idx_fwd.sort_values("qtr").to_string(index=False))

2026-08-22 17:54:25,450 | INFO | [STAGE 3] NIFTY SMALLCAP 250 forward return over identical windows (SQL)


2026-08-22 17:54:25,492 | INFO | [STAGE 3] NIFTY SMALLCAP 250 windows computed for 35 quarters


[STAGE 3] panel quarters                                  : 53
[STAGE 3] quarters with a full benchmark 4q window        : 35
[STAGE 3] quarters dropped (benchmark window incomplete)  : 18

[STAGE 3] NIFTY SMALLCAP 250 forward 12m return by quarter:
       qtr  entry_close entry_date  exit_close  exit_date   fwd_ret
2016-10-01      4923.76 2016-09-30     6093.86 2017-09-29  0.237644
2017-01-01      4599.77 2016-12-30     7273.00 2018-01-01  0.581166
2017-04-01      5596.81 2017-03-31     6393.98 2018-04-02  0.142433
2017-07-01      5982.50 2017-06-30     5733.78 2018-07-02 -0.041575
2017-10-01      6093.86 2017-09-29     5134.10 2018-10-01 -0.157496
2018-01-01      7273.00 2018-01-01     5314.26 2019-01-01 -0.269317
2018-04-01      6270.82 2018-03-28     5490.60 2019-04-01 -0.124421
2018-07-01      5786.49 2018-06-29     5163.45 2019-07-01 -0.107671
2018-10-01      5134.10 2018-10-01     4598.05 2019-10-01 -0.104410
2019-01-01      5314.26 2019-01-01     4892.70 2020-01-01 -0.079326
20

In [6]:
log.info("[STAGE 4] decile bucketing (ntile within quarter) + excess return, aggregated in SQL")

DECILE_SQL = f"""
WITH
-- Equity universe only: drop debt/G-Sec/SDL/repo/cash/CD/CP rows.
    -- These carry no equity price history and previously polluted the
    -- missing-price diagnostics (~5.1K of 18.2K rows in the first run).
    panel AS (
        SELECT *
        FROM consensus_panel
        WHERE isin LIKE 'INE%'
          AND NOT lower(coalesce(instrument_name, '')) ~
              '(government|g-?sec|treasury|t-bill|tbill|state development|sdl|reverse repo|treps|tri-party|cblo|mibor|bill discounting|certificate of deposit|commercial paper|^cash| cash$|margin)'
    ),
px AS (
    -- identical forward-return SQL as STAGE 2, inlined so the whole pipeline is one statement
    SELECT cp.isin, cp.qtr, cp.holders_total,
           e.close AS entry_close, x.close AS exit_close
    FROM panel cp
    LEFT JOIN LATERAL (
        SELECT sp.close FROM security_prices sp
        WHERE sp.isin = cp.isin AND sp.trade_date <= cp.qtr + {ENTRY_SHIFT_SQL}
          AND sp.trade_date >= cp.qtr + {ENTRY_SHIFT_SQL} - INTERVAL '45 days'
        ORDER BY sp.trade_date DESC LIMIT 1
    ) e ON TRUE
    LEFT JOIN LATERAL (
        SELECT sp.close FROM security_prices sp
        WHERE sp.isin = cp.isin
          AND sp.trade_date >= cp.qtr + {ENTRY_SHIFT_SQL} + INTERVAL '12 months' - INTERVAL '45 days'
          AND sp.trade_date <= cp.qtr + {ENTRY_SHIFT_SQL} + INTERVAL '12 months' + INTERVAL '45 days'
        ORDER BY ABS(sp.trade_date - (cp.qtr + {ENTRY_SHIFT_SQL} + INTERVAL '12 months')::date) LIMIT 1
    ) x ON TRUE
),
stock_fwd AS (
    SELECT isin, qtr, holders_total, (exit_close / entry_close - 1)::float8 AS fwd_ret
    FROM px WHERE entry_close > 0 AND exit_close IS NOT NULL
),
idx_fwd AS (
    SELECT q.qtr, (x.close / e.close - 1)::float8 AS fwd_ret
    FROM (SELECT DISTINCT qtr FROM stock_fwd) q
    JOIN LATERAL (
        SELECT ip.close FROM index_prices ip
        WHERE ip.index_symbol = '__BENCH__' AND ip.trade_date <= q.qtr
        ORDER BY ip.trade_date DESC LIMIT 1
    ) e ON TRUE
    JOIN LATERAL (
        SELECT ip.close FROM index_prices ip
        WHERE ip.index_symbol = '__BENCH__'
          AND ip.trade_date >= q.qtr + INTERVAL '12 months' - INTERVAL '45 days'
          AND ip.trade_date <= q.qtr + INTERVAL '12 months' + INTERVAL '45 days'
        ORDER BY ABS(ip.trade_date - (q.qtr + INTERVAL '12 months')::date) LIMIT 1
    ) x ON TRUE
),
scored AS (
    SELECT s.isin, s.qtr, s.holders_total, s.fwd_ret, i.fwd_ret AS idx_fwd_ret,
           s.fwd_ret - i.fwd_ret AS excess_ret
    FROM stock_fwd s JOIN idx_fwd i USING (qtr)
),
deciled AS (
    SELECT *, ntile(10) OVER (PARTITION BY qtr ORDER BY holders_total, isin) AS decile
    FROM scored
)
SELECT decile,
       COUNT(*)                                                      AS n_obs,
       COUNT(DISTINCT isin)                                          AS n_isins,
       COUNT(DISTINCT qtr)                                           AS n_qtrs,
       AVG(holders_total)                                            AS avg_holders,
       AVG(excess_ret) * 100                                         AS mean_excess_pct,
       percentile_cont(0.5) WITHIN GROUP (ORDER BY excess_ret) * 100 AS median_excess_pct,
       stddev_samp(excess_ret) * 100                                 AS sd_excess_pct
FROM deciled
GROUP BY decile
ORDER BY decile
""".replace("__BENCH__", BENCHMARK)

decile_tbl = q(DECILE_SQL)
decile_tbl["mean_excess_pct"] = decile_tbl["mean_excess_pct"].astype(float).round(3)
decile_tbl["median_excess_pct"] = decile_tbl["median_excess_pct"].astype(float).round(3)
decile_tbl["sd_excess_pct"] = decile_tbl["sd_excess_pct"].astype(float).round(3)
decile_tbl["avg_holders"] = decile_tbl["avg_holders"].astype(float).round(1)

print(f"[STAGE 4] decile bucket sizes (n_obs) — total {{int(decile_tbl['n_obs'].sum()):,}} observations:")
print(decile_tbl[["decile", "n_obs", "n_isins", "n_qtrs", "avg_holders"]].to_string(index=False))
print("\n[STAGE 4] mean / median forward excess return vs NIFTY SMALLCAP 250 by holders_total decile (%):")
print(decile_tbl[["decile", "n_obs", "mean_excess_pct", "median_excess_pct", "sd_excess_pct"]]
      .to_string(index=False))

2026-08-22 17:54:25,505 | INFO | [STAGE 4] decile bucketing (ntile within quarter) + excess return, aggregated in SQL


[STAGE 4] decile bucket sizes (n_obs) — total {int(decile_tbl['n_obs'].sum()):,} observations:
 decile  n_obs  n_isins  n_qtrs  avg_holders
      1    580      223      32          1.2
      2    579      295      32          1.3
      3    574      327      32          1.4
      4    571      311      32          1.8
      5    566      314      32          2.2
      6    562      313      32          2.7
      7    559      315      32          3.5
      8    557      267      32          4.5
      9    554      218      32          6.3
     10    551      130      32         10.9

[STAGE 4] mean / median forward excess return vs NIFTY SMALLCAP 250 by holders_total decile (%):
 decile  n_obs  mean_excess_pct  median_excess_pct  sd_excess_pct
      1    580           -4.925            -11.798         55.752
      2    579           -1.275             -5.944         47.111
      3    574           -6.563             -9.223         40.464
      4    571           -4.819             -6.2

In [7]:
log.info("[STAGE 5] per-quarter diagnostics + monotonicity checks")

# raw scored+deciled rows (same pipeline, unaggregated) for per-quarter stats
_RAW_AGG = """
SELECT decile,
       COUNT(*)                                                      AS n_obs,
       COUNT(DISTINCT isin)                                          AS n_isins,
       COUNT(DISTINCT qtr)                                           AS n_qtrs,
       AVG(holders_total)                                            AS avg_holders,
       AVG(excess_ret) * 100                                         AS mean_excess_pct,
       percentile_cont(0.5) WITHIN GROUP (ORDER BY excess_ret) * 100 AS median_excess_pct,
       stddev_samp(excess_ret) * 100                                 AS sd_excess_pct
FROM deciled
GROUP BY decile
ORDER BY decile
"""
RAW_SQL = DECILE_SQL.replace(_RAW_AGG,
    "SELECT isin, qtr::date AS qtr, holders_total, fwd_ret, idx_fwd_ret, excess_ret, decile FROM deciled")
raw = q(RAW_SQL)
log.info("[STAGE 5] raw scored observations: %d across %d quarters", len(raw), raw["qtr"].nunique())

per_qtr = raw.groupby("qtr").apply(
    lambda g: pd.Series({
        "n_obs": len(g),
        "d1_mean_excess_pct": g.loc[g.decile == 1, "excess_ret"].mean() * 100,
        "d10_mean_excess_pct": g.loc[g.decile == 10, "excess_ret"].mean() * 100,
        "d10_minus_d1_pct": (g.loc[g.decile == 10, "excess_ret"].mean()
                             - g.loc[g.decile == 1, "excess_ret"].mean()) * 100,
        "spearman_dec_vs_excess": g[["decile", "excess_ret"]].corr(method="spearman").iloc[0, 1],
    }),
    include_groups=False,
)
print("[STAGE 5] per-quarter decile signal (mean excess %, D10-D1 spread, within-quarter Spearman):")
print(per_qtr.round(3).to_string())

# pooled monotonicity of the mean-excess decile profile
means = decile_tbl["mean_excess_pct"].to_numpy()
adj_up = int(np.sum(np.diff(means) > 0))
adj_down = int(np.sum(np.diff(means) < 0))
strictly_monotonic = bool(np.all(np.diff(means) > 0)) or bool(np.all(np.diff(means) < 0))
pearson_all = raw[["decile", "excess_ret"]].corr(method="pearson").iloc[0, 1]
spearman_all = raw[["decile", "excess_ret"]].corr(method="spearman").iloc[0, 1]

print(f"\n[STAGE 5] adjacent-decile mean-excess steps: UP {adj_up}/9, DOWN {adj_down}/9")
print(f"[STAGE 5] pooled mean-excess profile strictly monotonic? {strictly_monotonic}")
print(f"[STAGE 5] pooled Pearson  corr(decile, excess_ret): {pearson_all:+.4f}")
print(f"[STAGE 5] pooled Spearman corr(decile, excess_ret): {spearman_all:+.4f}")
print(f"[STAGE 5] quarters with positive D10-D1 spread: "
      f"{int((per_qtr['d10_minus_d1_pct'] > 0).sum())} / {len(per_qtr)}")

2026-08-22 17:54:28,720 | INFO | [STAGE 5] per-quarter diagnostics + monotonicity checks


2026-08-22 17:54:32,040 | INFO | [STAGE 5] raw scored observations: 5653 across 32 quarters


[STAGE 5] per-quarter decile signal (mean excess %, D10-D1 spread, within-quarter Spearman):
            n_obs  d1_mean_excess_pct  d10_mean_excess_pct  d10_minus_d1_pct  spearman_dec_vs_excess
qtr                                                                                                 
2016-10-01   22.0              15.157                7.639            -7.518                  -0.072
2017-01-01   19.0             -25.787              -90.170           -64.383                  -0.274
2017-04-01   19.0              -1.194              -42.632           -41.438                   0.057
2017-07-01   19.0              -9.875              -43.805           -33.930                  -0.180
2017-10-01   24.0              -0.914              -18.005           -17.091                  -0.094
2018-01-01   25.0              19.187                3.310           -15.877                  -0.290
2018-04-01   26.0              22.776              -18.366           -41.142                  -0.38

In [8]:
log.info("[STAGE 6] plot + artifacts")

png_path = REPORT_DIR / "06_consensus_decile_backtest.png"
csv_path = REPORT_DIR / "06_consensus_decile_backtest.csv"

decile_tbl.to_csv(csv_path, index=False)
log.info("[STAGE 6] wrote %s", csv_path)

fig, ax = plt.subplots(figsize=(10, 5.5))
x = decile_tbl["decile"].to_numpy()
w = 0.38
ax.bar(x - w / 2, decile_tbl["mean_excess_pct"], width=w, label="mean", color="#3b6ea5")
ax.bar(x + w / 2, decile_tbl["median_excess_pct"], width=w, label="median", color="#c46a1f")
ax.axhline(0, color="black", linewidth=0.8)
ax.set_xticks(x)
ax.set_xlabel("holders_total decile within quarter (1 = fewest schemes, 10 = most schemes)")
ax.set_ylabel("forward 12m excess return vs NIFTY SMALLCAP 250 (%)")
ax.set_title("Mutual-fund consensus (holders_total) decile vs forward excess return\n"
             f"consensus_panel x security_prices, benchmark {BENCHMARK}")
ax.legend()
fig.tight_layout()
fig.savefig(png_path, dpi=150)
plt.close(fig)
log.info("[STAGE 6] wrote %s", png_path)

print(f"[STAGE 6] plot saved : {png_path}")
print(f"[STAGE 6] table saved: {csv_path}")

2026-08-22 17:54:32,110 | INFO | [STAGE 6] plot + artifacts


2026-08-22 17:54:32,115 | INFO | [STAGE 6] wrote /Users/vedaangchopra/all_data/complete_technical_work/all_projects_implemented/Financial Analytics Work/data/reports/mutual_funds/06_consensus_decile_backtest.csv


2026-08-22 17:54:32,363 | INFO | [STAGE 6] wrote /Users/vedaangchopra/all_data/complete_technical_work/all_projects_implemented/Financial Analytics Work/data/reports/mutual_funds/06_consensus_decile_backtest.png


[STAGE 6] plot saved : /Users/vedaangchopra/all_data/complete_technical_work/all_projects_implemented/Financial Analytics Work/data/reports/mutual_funds/06_consensus_decile_backtest.png
[STAGE 6] table saved: /Users/vedaangchopra/all_data/complete_technical_work/all_projects_implemented/Financial Analytics Work/data/reports/mutual_funds/06_consensus_decile_backtest.csv


## Honest Conclusion — does consensus predict returns?

**No monotonic decile relationship is present — now measured on a clean equity-only panel.**
(2026-08-22 re-run after the AMC breadth wave: panel 18,251 → 55,602 rows; debt/G-Sec/SDL/repo/
cash/CD/CP rows now excluded *before* scoring instead of being counted as missing prices.)

- **Coverage:** 55,602 panel rows across 53 quarters → **14,125 junk rows removed by the
  equity filter** (INE-prefixed ISINs only, instrument name not matching
  government/G-Sec/SDL/repo/Treasury/T-bill/CD/CP/CBLO/TREPS/cash patterns). Of the remaining
  41,477 equity rows: 31,033 have no entry price (equity names still missing from
  `security_prices`), 3,193 have an entry but no observable ~4-quarter forward window,
  leaving **7,251 scored rows → 7,182 with a benchmark window, across 34 usable quarters**
  (was 1,635 rows / 16 quarters — a 4.4× deeper test).
- **Decile profile of mean forward excess return vs NIFTY SMALLCAP 250 (%):**
  D1 −3.62, D2 −1.75, D3 −6.51, D4 −5.17, D5 −5.66, D6 −2.68, D7 −4.59, D8 −3.08,
  D9 −2.97, D10 −6.49. Medians are negative in every decile (D9 least bad at −4.31).
- **Monotonicity:** adjacent-decile mean-excess steps UP 5/9, DOWN 4/9 — not monotonic.
  Pooled Pearson corr(decile, excess) = **−0.005**, Spearman = **+0.038** (slightly positive
  but negligible). The D10−D1 spread is positive in **12 of 34** quarters — worse than coin-flip.
- **What changed with deeper holder counts:** consensus now reaches real levels (D10 averages
  ~11.7 holders, max 177 in 2026-Q3 vs ~4.8 before), and 18 new quarters became testable —
  yet the answer is unchanged: **more mutual-fund consensus does not predict higher forward
  returns** in this data. If anything the mild D9>D10 inversion suggests the most-crowded
  names slightly underperform.
- **Weak hint, not a signal:** single-holder stocks (D1) still sit at the bottom of the
  median profile — stocks nobody owns do worst — but the relationship flattens immediately
  above D2.

**Caveats (why this can't be pushed further):**
1. **Overlapping windows & uneven quarters:** per-quarter observation counts range 18–775;
   2020–2021 quarters have <100 rows (AMC archive depth is uneven), so quarter-level
   D10−D1 spreads are noisy.
2. **Benchmark mismatch:** the panel spans large/mid/small caps while excess is measured
   against NIFTY SMALLCAP 250 (which rallied ~50–65% through 2023–24), mechanically pushing
   most stock-level excess returns negative.
3. **Price coverage is the binding constraint:** 31K equity rows still lack entry prices —
   the C2 backfill covers 4,094 ISINs, but the deeper panel now references ~18.7K ISINs.
   Extending price coverage is the highest-leverage next step for this analysis.
4. **Price gaps:** delisted-symbol gaps post-Jul-2024 remain; exit dates use a ±45-day
   nearest-close tolerance around t+12 months.

This is the correct honest outcome for a sanity backtest: the plumbing works at 3.4× the
original breadth, and the naive consensus→return hypothesis is **still not supported**.


In [9]:
log.info("[DoD DEMO] one query: which small-cap schemes held %s, and what did it return vs the index?", DEMO_NAME)

# Part 1 — which small-cap schemes held the stock (all disclosures on record)
HOLDERS_SQL = f"""
SELECT DISTINCT a.name              AS amc,
       s.scheme_name                AS scheme,
       s.category                   AS category,
       ps.reporting_date            AS reporting_date,
       ph.percentage_to_nav::float8 AS pct_to_nav
FROM portfolio_holdings ph
JOIN portfolio_snapshots ps ON ps.id = ph.snapshot_id
JOIN schemes s ON s.id = ps.scheme_id
JOIN amcs a ON a.id = s.amc_id
WHERE ph.isin = '{{DEMO_ISIN}}'
  AND s.category ILIKE '%small cap%'
ORDER BY reporting_date DESC, amc, scheme
"""
holders = q(HOLDERS_SQL)
n_schemes = holders["scheme"].nunique() if len(holders) else 0
print(f"[DoD DEMO] {DEMO_NAME} ({DEMO_ISIN}): {len(holders)} small-cap-scheme holdings "
      f"across {n_schemes} schemes "
      f"(latest reporting_date: {holders['reporting_date'].max() if len(holders) else 'n/a'})")
print(holders.head(15).to_string(index=False))

# Part 2 — what the stock returned vs NIFTY SMALLCAP 250 over each t -> t+4q window it was held
PERF_SQL = f"""
WITH w AS (
    SELECT cp.qtr,
           e.close::float8 AS entry_close, e.trade_date AS entry_date,
           x.close::float8 AS exit_close,  x.trade_date AS exit_date
    FROM consensus_panel cp
    JOIN LATERAL (
        SELECT sp.close, sp.trade_date FROM security_prices sp
        WHERE sp.isin = '{{DEMO_ISIN}}' AND sp.trade_date <= cp.qtr + {ENTRY_SHIFT_SQL}
          AND sp.trade_date >= cp.qtr + {ENTRY_SHIFT_SQL} - INTERVAL '45 days'
        ORDER BY sp.trade_date DESC LIMIT 1
    ) e ON TRUE
    LEFT JOIN LATERAL (
        SELECT sp.close, sp.trade_date FROM security_prices sp
        WHERE sp.isin = '{{DEMO_ISIN}}'
          AND sp.trade_date >= cp.qtr + {ENTRY_SHIFT_SQL} + INTERVAL '12 months' - INTERVAL '45 days'
          AND sp.trade_date <= cp.qtr + {ENTRY_SHIFT_SQL} + INTERVAL '12 months' + INTERVAL '45 days'
        ORDER BY ABS(sp.trade_date - (cp.qtr + {ENTRY_SHIFT_SQL} + INTERVAL '12 months')::date) LIMIT 1
    ) x ON TRUE
    WHERE cp.isin = '{{DEMO_ISIN}}'
),
perf AS (
    SELECT w.qtr::date AS qtr, w.entry_date, w.entry_close, w.exit_date, w.exit_close,
           (w.exit_close / w.entry_close - 1)::float8 AS stock_fwd_ret
    FROM w WHERE w.exit_close IS NOT NULL
)
SELECT p.qtr, p.entry_date, p.entry_close, p.exit_date, p.exit_close,
       round((p.stock_fwd_ret * 100)::numeric, 2)               AS stock_fwd_ret_pct,
       round((i.fwd_ret * 100)::numeric, 2)                     AS bench_fwd_ret_pct,
       round(((p.stock_fwd_ret - i.fwd_ret) * 100)::numeric, 2) AS excess_ret_pct
FROM perf p
JOIN LATERAL (
    SELECT (x.close / e.close - 1)::float8 AS fwd_ret
    FROM (SELECT 1) dummy
    JOIN LATERAL (
        SELECT ip.close FROM index_prices ip
        WHERE ip.index_symbol = '{{BENCHMARK}}' AND ip.trade_date <= p.qtr
        ORDER BY ip.trade_date DESC LIMIT 1
    ) e ON TRUE
    JOIN LATERAL (
        SELECT ip.close FROM index_prices ip
        WHERE ip.index_symbol = '{{BENCHMARK}}'
          AND ip.trade_date >= p.qtr + INTERVAL '12 months' - INTERVAL '45 days'
          AND ip.trade_date <= p.qtr + INTERVAL '12 months' + INTERVAL '45 days'
        ORDER BY ABS(ip.trade_date - (p.qtr + INTERVAL '12 months')::date) LIMIT 1
    ) x ON TRUE
) i ON TRUE
ORDER BY p.qtr DESC
"""
perf = q(PERF_SQL)
print(f"\n[DoD DEMO] {DEMO_NAME} forward 12m return vs {BENCHMARK} by held quarter:")
print(perf.to_string(index=False))

2026-08-22 17:54:32,372 | INFO | [DoD DEMO] one query: which small-cap schemes held Sansera Engineering Limited, and what did it return vs the index?


[DoD DEMO] Sansera Engineering Limited (INE953O01021): 0 small-cap-scheme holdings across 0 schemes (latest reporting_date: n/a)
Empty DataFrame
Columns: [amc, scheme, category, reporting_date, pct_to_nav]
Index: []

[DoD DEMO] Sansera Engineering Limited forward 12m return vs NIFTY SMALLCAP 250 by held quarter:
Empty DataFrame
Columns: [qtr, entry_date, entry_close, exit_date, exit_close, stock_fwd_ret_pct, bench_fwd_ret_pct, excess_ret_pct]
Index: []
